### DISCUSSION TITLE
Copilot in Visual Studio Formatting Incorrect (new lines + extra tabs)

### DISCUSSION THREAD

**Discussion by CommanderFoo** [community/community#60171] [Copilot Conversations] [answered]
### Select Topic Area

Bug

### Body

Hello,

In Visual Studio 17.6.4 using Copilot 1.94.0.0, any code (C#) that is automatically inserted is never formatted correctly, it will insert extra tabs and new lines. Testing this in Visual Code behaves as expected, so it seems to be a bug that some people are still experiencing.

See the suggestion vs the inserted code. This is also happening for my other team members.

![image](https://github.com/community/community/assets/944078/508f91f1-cd1e-4fa7-b867-80bf982b4b78)
![image](https://github.com/community/community/assets/944078/6f7ab0b1-cfbc-490b-bda4-f35c7d55206c)

**Comment by luapi82** [comment]
This has also been plaguing me ever since I started using GitHub Copilot with Visual Studio 2022 and it still and issue for me as well. I hope this gets fixed soon as it is quite inefficient.

**Reply by ayim** [reply]
@luapi82 hello from Copilot for Visual Studio team - logging a bug for this and we're looking at root causing this issue.
Feel free to reach out to aayim@microsoft.com, we'd love to have a 15 min chat about any other bugs or behaviour you're encountering.

**Reply by Mathetis** [reply]
Also seeing this problem consistently. VS 2022, C#, Visual Assist is installed.

**Comment by michaellperry** [comment]
Here's an example from VS 2022 version 17.6.5, GitHub Copilot agent version 1.94.220:

Suggestion:
![image](https://github.com/community/community/assets/68170/8a21ffa6-86b1-4a02-8312-8045aeda936e)

Inserted code:
![image](https://github.com/community/community/assets/68170/8b96626e-6e9d-477a-899b-a610797ad53c)

It appears that Visual Studio accepts the Enter key and indents the next line. Then CoPilot suggests further indentation, assuming that the line was not already indented.

**Reply by ayim** [reply]
@michaellperry hello from Copilot for Visual Studio team - logging a bug for this and we're looking at root causing this issue.
Interesting that this is happening in SQL as well.
Feel free to reach out to aayim@microsoft.com, we'd love to have a 15 min chat about any other bugs or behaviour you're encountering.

**Comment by trevorkerby** [comment]
I'm also experiencing this bug, with the exact behavior described by OP; the suggestion is formatted just fine, but once it's applied then it becomes a disaster. 

With this bug, many suggestions that could have saved time are now affecting my flow because I have to go back and untangle an indentation hell. It would be awesome if this got some attention, because it really makes using the product feel like a poor experience.

**Reply by ayim** [reply]
@trevorkerby hello from Copilot for Visual Studio team - logging a bug for this and we're looking at root causing this issue.
Just to confirm, you're seeing this in C#?
Feel free to reach out to aayim@microsoft.com, we'd love to have a 15 min chat about any other bugs or behaviour you're encountering.

**Reply by LockTar** [reply]
@ayim I can confirm that I have it in C#

**Reply by ayim** [reply]
@LockTar working on the fix, tracking here: https://developercommunity.visualstudio.com/t/Extra-spaces-before-the-method-parameter/10517858

**Comment by sterlingstokes** [comment]
I was experiencing this for quite a while, also. As is the case with a lot of current issues around AI, I was able to resolve this issue by just tweaking my prompt to say things like "GENERATE THIS CODE FROM THIS POINT FORWARD" or "Replace these contents -- DO NOT INSERT CODE HERE".

**Reply by ayim** [reply]
Prompt engineering FTW huh? That's awesome, I'm going to start using this.

**Comment by froodybanana** [comment]
As far as I can tell this happens because when you hit Enter in Visual Studio, the indentation doesn't actually exist until you type something.

So the cursor is not really where it it's shown at.

If you hit space then backspace, it creates the indentation, and the suggestion goes where it's supposed to.

No idea if this is a VS issue or a Copilot issue, but I'd imagine VS is the real culprit here.

**Reply by ayim** [reply]
@mtnielsen interesting, so this occurs on ACCEPT.
Looks like it's a formatting issue in some languages - are you in C#? What languages do you see this issue in?
Feel free to reach out to [aayim@microsoft.com](mailto:aayim@microsoft.com), we'd love to have a 15 min chat about any other bugs or behaviour you're encountering.

**Reply by froodybanana** [reply]
You may have sneak-fixed it in C# without my noticing, but I still get it when writing Razor pages.

I'm not sure if this shows up as a fancy video, but I did a quick recording

https://github.com/community/community/assets/5136850/f994a552-c790-4619-8ec2-23f46b6439fa

**Reply by ayim** [reply]
No fix yet.... tracking the issue here: https://developercommunity.visualstudio.com/t/Extra-spaces-before-the-method-parameter/10517858

**Comment by bjornharink** [comment]
Same issue here using it to create function documentation. Even changed to right arrow to see if it fixes the issue.

Using Visual Studio Community 17.8.1, Copilot 1.137.0.0

![image](https://github.com/community/community/assets/13664890/10124b14-b6cd-4b7f-aae6-6cf09e4f4cec)
![image](https://github.com/community/community/assets/13664890/42d7ac43-4a87-405c-a640-5bf1a996560e)

**Reply by ayim** [reply]
@bjornharink just to confirm, this looks like C++? Any other languages you see this occur in?

**Reply by bjornharink** [reply]
This is C/C++ (embedded). Currently, using only that in Visual Studio.

**Reply by ayim** [reply, marked as answer]
Tracking the fix here: https://developercommunity.visualstudio.com/t/Extra-spaces-before-the-method-parameter/10517858

**Comment by EvoTechMike** [comment]
With latest update it still happens

![image](https://github.com/community/community/assets/134078277/3bd65bb1-40bd-4cab-a31a-196477b2ead5)

![image](https://github.com/community/community/assets/134078277/3d3f6def-56cb-4367-a4f8-7f9b02b76f41)

It doesn't like constructors

**Reply by ayim** [reply]
Interesting - could you file a bug report via Report a Problem?
We implemented a fix but need to investigate further since it's still happening to ya.
![image](https://github.com/community/community/assets/10951869/104cd0ef-9871-4398-aaf3-e8c6a9a8fc98)

Send me the issue link once it's filed and i'll make sure to expedite.

**Reply by aaron-cui-sh** [reply]
Same here, updated to visual studio 17.9.7

**Comment by ultrabstrong** [comment]
I am also still experiencing this issue on Version 17.9.7

**Comment by noellelc** [comment]
This has been fixed in Version 17.10, which is now available. Please update to the latest version to take advantage of this fix.

**Reply by ultrabstrong** [reply]
I was able to confirm 17.10 fixed the issue for me

**Comment by kasgithub8** [comment]
Hello,
I'm on Visual Studio 2022 Version 17.13.6 and have had this issue for a couple of weeks now.
When I write C# in .razor files it looks like this:
```               
                @if (showOnlyActiveData)
                {

}
else
                        {

                        }
```

And in the @code portion of the .razor file:
```
private void TestMethod()
    {

    }
```

Does anybody share this problem?

**Reply by noellelc** [reply]
Please update to the latest version of VS (as of today, for VS 2022, that's 17.14.24 and VS 2026 18.2.0).

**Reply by kasgithub8** [reply]
Thanks @noellelc,
I believe the issue is fixed but it's still popping in here and there, I can't pinpoint when and why tho.
But thank you, it's way better now!

**Comment by Kocaak** [comment]
Hello,
I am using version 1.108 and this problem sis till ongoing for me. Despite the spacing looking correct in IPython notebooks, I always get "inconsistent use of tabs and spaces in indentation" in copilot generated code.

**Reply by noellelc** [reply]
Please update to the latest version of VS (as of today, for VS 2022, that's 17.14.24 and VS 2026 18.2.0). Version 1.108 suggests you may be on a much older version of VS where Copilot was not directly integrated.

**Reply by Kocaak** [reply]
Thank you for your swift reply. I am using VS Code (sorry if it wasn't clear), and the latest is 1.108. Please let me know if I am missing anything.

**Reply by noellelc** [reply]
This post is specifically for the Visual Studio IDE. While we are sister products, we are different development teams. You'll want to reach out to VS Code through their feedback channels.
